dashboard/app/models/pd/workshop_material_order.rb

```diff
@@ -0,0 +1,235 @@
+      [street, apartment_or_suite, city, state, zip_code].compact.join(', ')
+    end
+
+    def address_fields_changed?
+      street_changed? || apartment_or_suite_changed? || city_changed? || state_changed? || zip_code_changed?
+    end
+
+    def expected_ship_date
+      return nil unless order_response
+      MimeoRestClient.parse_date order_response['RecipientDetails'].first['ExpectedShipDate']
+    end
+
+    def expected_delivery_date
+      return nil unless order_response
+      MimeoRestClient.parse_date order_response['RecipientDetails'].first['ExpectedDeliveryDate']
+    end
+
+    def order_response
+      raw = read_attribute :order_response
+      raw.nil? ? nil : JSON.parse(raw)
+    end
+
+    def order_error
+      raw = read_attribute :order_error
+      raw.nil? ? nil : JSON.parse(raw)
+    end
+
+    def ordered?
+      ordered_at.present?
+    end
+
+    def shipped?
+      order_status == 'Shipped'
+    end
+
+    # Place order with the Mimeo API, based on shipping info in a valid model
+    # Save order_attempted_at to the DB, and
+    #   on success: ordered_at, order_response, order_id
+    #   on error: order_error
+    # This is idempotent. Once an order has been successfully placed, subsequent calls will
+    #   return the existing response.
+    # @return [Hash, nil] either the raw order response, or nil if no order was made or it failed
+    # @raise [RuntimeError] if the model fails validation.
+    def place_order
+      raise "Fix errors before ordering: #{errors.full_messages}" unless valid?
+      return order_response if ordered?
+
+      self.order_attempted_at = Time.zone.now
+      response = nil
+      begin
+        response = MimeoRestClient.new.place_order(
+          first_name: enrollment.first_name,
+          last_name: enrollment.last_name,
+          company_name: school_or_company,
+          street: street,
+          apartment_or_suite: apartment_or_suite,
+          city: city,
+          state_or_province: state,
+          country: 'US',
+          postal_code: zip_code,
+          email: enrollment.email,
+          phone_number: phone_number
+        )
+
+        self.ordered_at = order_attempted_at
+        self.order_response = response.to_json
+        self.order_error = nil
+        self.order_id = response['OrderFriendlyId']
+      rescue RestClient::ExceptionWithResponse => error
+        self.order_error = report_error(:place_order, error).to_json
+      ensure
+        save!
+      end
+
+      response
+    end
+
+    # Check the status of an order with the Mimeo API
+    # Save order_status_last_checked_at in the DB, and when the status changes:
+    #   order_status and order_status_changed_at
+    # @return [String, nil] last known order status string, or nil if no order has been made
+    # See MimeoRestClient::STATUS for available status strings.
+    # Once 'Shipped', it will not be checked again.
+    def check_status
+      return nil unless order_id
```

Comment:
nil stands in for NO_ORDER. Is there any reason to have an additional value for that?

I don't want to raise an exception here, so the end user doesn't see any errors. Note we do log to honeybadger when an API call fails unexpectedly.